### PR TITLE
Range Locking: Allow different LockManagers, add Range Lock definitions

### PR DESCRIPTION
**Pull request by spetrunia**
This PR has two commits:
1.  Modify the code to allow different Lock Managers (of any kind) to be used.  It is implied that a LockManager uses its own custom LockTracker.
2.  Add definitions for Range Locking (class Endpoint and GetRangeLock() function.

@cheng-chang, is this what you've had in mind (should the PR have both item 1 and item 2?)

### CODE REVIEW

File: include/rocksdb/utilities/transaction_db.h
Comment (cheng-chang) [inline include/rocksdb/utilities/transaction_db.h:34]: nit: can we use "Manager" consistently?

File: utilities/transactions/lock/lock_mgr.h
Comment (cheng-chang) [inline utilities/transactions/lock/lock_mgr.h:46]: Is a `LockTrackerFactory` `new`ed each time when calling `getLockTrackerFactory`? If so, can we call it `NewLockTrackerFactory` and leave a note indicating that the caller owns the returned factory?

Comment (cheng-chang) [reply utilities/transactions/lock/lock_mgr.h:46]: Seems like a static instance is returned each time, then can we comment that the caller does not own the returned object?

Comment (cheng-chang) [reply utilities/transactions/lock/lock_mgr.h:46]: nit: `Get` for consistency?

File: utilities/transactions/lock/lock_mgr.h
Comment (cheng-chang) [inline utilities/transactions/lock/lock_mgr.h:63]: nit: capitalize `Init` for consistency?

File: utilities/transactions/lock/point/point_lock_tracker.h
Comment (cheng-chang) [inline utilities/transactions/lock/point/point_lock_tracker.h:1]: I assume this file is only moved from `lock/lock_tracker.h` without modification?

File: utilities/transactions/lock/point/point_lock_tracker.cc
Comment (cheng-chang) [inline utilities/transactions/lock/point/point_lock_tracker.cc:1]: I assume this file is only moved from `lock/lock_tracker.cc` without modification?

File: utilities/transactions/lock/lock_mgr.h
Comment (cheng-chang) [inline utilities/transactions/lock/lock_mgr.h:51]: can we consistently use the `ColumnFamilyId` typedef instead of `uint32_t`? also applies to other occurrences of cf id.

File: utilities/transactions/lock/point/point_lock_tracker.h
Comment (cheng-chang) [inline utilities/transactions/lock/point/point_lock_tracker.h:89]: can we make this private and have a public `Get` method instead?

File: include/rocksdb/utilities/transaction_db.h
Comment (cheng-chang) [inline include/rocksdb/utilities/transaction_db.h:59]: this feels like specific to certain range lock implementations?

File: utilities/transactions/lock/point/point_lock_mgr.h
Comment (cheng-chang) [inline utilities/transactions/lock/point/point_lock_mgr.h:41]: can we rename it to `PointLockManager` to be consistent with namings like `PointLockTracker`?

File: utilities/transactions/pessimistic_transaction.h
Comment (cheng-chang) [inline utilities/transactions/pessimistic_transaction.h:122]: Ideally, there will be a range lock manager interface, and the implementation here will be based on that interface. 

In current design, will this method's implementation be bounded to a specific implementation of range lock?

File: utilities/transactions/pessimistic_transaction_db.cc
Comment (cheng-chang) [inline utilities/transactions/pessimistic_transaction_db.cc:45]: would `reset(mgr)` work here?

**Review by cheng-chang — COMMENTED** [review]
Thanks @spetrunia for this PR. 

The current `RangeLockManagerHandle` looks too general to me. Although we have `BaseLockMgr` interface, we don't have an interface for `RangeLockMgr`, and the `LockRange` method in `Transaction` will depend on the implementation of range lock.

I'm thinking can we have a general `RangeLockManager` interface, and implement `Transaction::LockRange` with that interface. Then in https://github.com/facebook/rocksdb/pull/7506, we just implement the interface with toku's range tree?

**Comment by cheng-chang** [comment]
I tried to abstract out the LockManager interface at https://github.com/facebook/rocksdb/pull/7532, but that's just the first step, when implementing the range lock, updates to the interface may be needed.

File: include/rocksdb/utilities/transaction_db.h
Comment (cheng-chang) [inline include/rocksdb/utilities/transaction_db.h:62]: Are the concept of lock memory and escalation count are specific for the range lock implementation based on range tree?

File: utilities/transactions/lock/lock_mgr.h
Comment (cheng-chang) [inline utilities/transactions/lock/lock_mgr.h:44]: This interface feels like for point lock only? E.g. there is no API like `TryLockRange`?

File: include/rocksdb/utilities/transaction_db.h
Comment (spetrunia) [reply include/rocksdb/utilities/transaction_db.h:62]: `current_lock_memory` is the total amount of memory currently used for lock data structures. It is applicable even to Point Lock Manager (although the Point Lock Manager doesn't report the number currently). 

for `escalation_count`... it may depend. One of the goals of the Range-based Lock Manager is that it can naturally support  Lock Escalation, so I guess the answer is more "yes" than "no".   One can imagine a range-based Lock Manager that doesn't support lock escalation but it won't be as useful.

File: utilities/transactions/lock/lock_mgr.h
Comment (spetrunia) [reply utilities/transactions/lock/lock_mgr.h:44]: Yes. 

Yes. The reasons for this is that class `PessimisticTransactionDB` has several places where it "knows" which kind of lock manager it is using and so it calls to `RangeLockMgr` : 

```c++
void PessimisticTransactionDB::init_lock_manager() {
    range_lock_mgr_ = lock_mgr_->getRangeLockMgr();
```

```c++
Status PessimisticTransactionDB::TryRangeLock(PessimisticTransaction* txn,
                                              uint32_t cfh_id,
                                              const Endpoint& start_endp,
                                              const Endpoint& end_endp) {
  if (range_lock_mgr_) {
    return range_lock_mgr_->TryRangeLock(txn, cfh_id, start_endp, end_endp,
                                         /*exclusive=*/true);
  } else
    return Status::NotSupported();
}
```

```c++
void PessimisticTransactionDB::UnLock(PessimisticTransaction* txn,
                                      const LockTracker& keys,
                                      bool all_keys_hint) {
  if (all_keys_hint && range_lock_mgr_) {
    range_lock_mgr_->UnLockAll(txn, GetEnv());
    return;
  }
  lock_mgr_->UnLock(txn, keys, GetEnv());
}
```

Should this code be changed so that the calls are made through the `BaseLockMgr`  pointer and all difference is hidden in the lock manager implementation?

**Comment by spetrunia** [comment]
> I'm thinking can we have a general `RangeLockManager` interface, and implement `Transaction::LockRange` with that interface. Then in #7506, we just implement the interface with toku's range tree?

This looks like something I could try doing. 
Which PR should this change go into?

**Comment by spetrunia** [comment]
> The current `RangeLockManagerHandle` looks too general to me.

The idea of this object was that it is visible to the RocksDB's user, so it was intentionally very opaque.

The user is supposed to be able to:
* Create a lock manager with `NewRangeLockManager`, and just get an opaque handle 
* Pass the Handle to RocksDB in `TransactionDBOptions::lock_mgr_handle` .
* When the DB is running, use the handle for accessing the user-visible knobs of the lock manager - call `set_max_lock_memory()` or get counters with `GetStatus()`.

`RangeLockMgrHandle` achieves all of the above.


If we get rid of it,  what alternative solution should we use?
(I'm looking at how other Lock Manager-related data is exposed to the user...  There are pure virtual methods defined in TransactionDB()  (e.g. `GetDeadlockInfoBuffer`).  Then there is an implementation in PessimisticTransactionDB, which passes the call to the Lock Manager.   Is this the pattern to follow?)

**Comment by cheng-chang** [comment]
@spetrunia https://github.com/facebook/rocksdb/pull/7532 is merged. Can you merge this PR with https://github.com/facebook/rocksdb/pull/7506, and ideally put the new files under the `transactions/lock/range/range_tree` directory?

File: utilities/transactions/lock/lock_manager.cc
Comment (cheng-chang) [inline utilities/transactions/lock/lock_manager.cc:20]: `mgr` can be retrieved from `lock_mgr_handle`, why need to pass both as parameters?

**Review by cheng-chang — APPROVED** [review]
@spetrunia Thanks for the update! I only have one question left.

File: include/rocksdb/utilities/transaction.h
Comment (cheng-chang) [inline include/rocksdb/utilities/transaction.h:92]: `explicit`?

File: include/rocksdb/utilities/transaction.h
Comment (cheng-chang) [inline include/rocksdb/utilities/transaction.h:95]: `explicit`?

File: include/rocksdb/utilities/transaction_db.h
Comment (cheng-chang) [inline include/rocksdb/utilities/transaction_db.h:115]: remove `;`?

File: utilities/transactions/lock/point/point_lock_manager_test.cc
Comment (cheng-chang) [inline utilities/transactions/lock/point/point_lock_manager_test.cc:8]: the include path needs to be relative to rocksdb's root directory such as `utilities/transactions/lock/point/point_lock_manager_test.h`.

**Comment by cheng-chang** [comment]
@spetrunia Thanks for the update, it mostly looks good to me except some minor comments. The description of the PR seems outdated, can you update it? After merging this PR, can you also update https://github.com/facebook/rocksdb/pull/7506 by merging the master branch?

File: include/rocksdb/utilities/transaction_db.h
Comment (spetrunia) [reply include/rocksdb/utilities/transaction_db.h:115]: WIll do

File: include/rocksdb/utilities/transaction.h
Comment (spetrunia) [reply include/rocksdb/utilities/transaction.h:95]: Will do.

File: include/rocksdb/utilities/transaction.h
Comment (spetrunia) [reply include/rocksdb/utilities/transaction.h:92]: Will do.

File: utilities/transactions/lock/point/point_lock_manager_test.cc
Comment (spetrunia) [reply utilities/transactions/lock/point/point_lock_manager_test.cc:8]: Will do.

File: utilities/transactions/lock/lock_manager.cc
Comment (spetrunia) [reply utilities/transactions/lock/lock_manager.cc:20]: > mgr can be retrieved from lock_mgr_handle, why need to pass both as parameters?

This is because of the following:

`opt.lock_mgr_handle` is defined as:
```
  std::shared_ptr<LockManagerHandle> lock_mgr_handle;
```
while this function's return value is:
```
  std::shared_ptr<LockManager>
```

Classes `LockManager` and `LockManagerHandle` do not inherit from one another (one needs to call getLockManager() to get one from the other).

And apparently, std::shared_ptr has a special constructor template for such cases, called "aliasing constructor":

```
return std::shared_ptr<LockManager>(opt.lock_mgr_handle, mgr);
```
this produces a shared pointer looks like a `LockManager*` when dereferenced, but its free-on-delete part shares the pointer to `LockManagerHandle` so it can share with `opt.lock_mgr_handle`. 

(I'm sorry if I'm re-stating the obvious here... It's the first time I saw/used this constructor).

**Comment by cheng-chang** [comment]
The format check fails, can you run `make format`? I think we are good to go after formatting.

**Pull request closed without merging**
